File: leetcodepractice/1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.js
/**
 * @param {string} str1
 * @param {string} str2
 * @return {string}
 */
var gcdOfStrings = function(str1, str2) {
    const length1 = str1.length;
    const length2 = str2.length;

    const isDivisor = (l) => {
        if (length1 % l || length2 % l) {
            return false;
        }

        let [f1, f2] = [Math.floor(length1 / l), Math.floor(length2 / l)];
        
        return (
            str1.slice(0,l).repeat(f1) == str1 && 
            str1.slice(0,l).repeat(f2) == str2
        );
    };

    for (let l = Math.min(length1, length2); l > 0; l--){
        if (isDivisor(l)) {
            return str1.slice(0,l);
        }
    };

    return "";
};

File: leetcodepractice/1146-greatest-common-divisor-of-strings/greatest-common-divisor-of-strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        str1_length, str2_length = len(str1), len(str2)

        def isDivisor(l):
            if str1_length % l or str2_length % l:
                return False
            f1, f2 = str1_length // l, str2_length // l
            return str1[:l] * f1 == str1 and str1[:l] * f2 == str2

        for l in range(min(str1_length, str2_length), 0, -1):
            if isDivisor(l):
                return str1[:l]
        return ""
